discoverd: in-memory heartbeating

benbjohnson opened the pull request:
## Overview

This pull request moves the heartbeats out of the raft write path and persists them only in-memory on the discoverd leader. The leader is responsible for expiring service instances that do not heartbeat within their TTL.

Leaders cannot begin expiring instances until after they have had leadership for at least 2x TTL.


### Other notes

The vendored version of Bolt was really old and included a bug that caused a panic when shutting down Raft prematurely. It wasn't a safety issue but the panic was showing up in tests and was causing problems so I upgraded to the latest commit.

Closes #2158
Closes #2168
Closes #2073

titanous commented:
Please split the bolt update into a separate commit.

titanous commented on discoverd/server/store.go:
```diff
@@ -475,20 +532,20 @@ func (s *Store) applyRemoveInstanceCommand(cmd []byte) error {
 	// Remove instance data.
-	entry := s.data.Instances[c.Service][c.ID]
+	inst := s.data.Instances[c.Service][c.ID]
 	delete(s.data.Instances[c.Service], c.ID)
```
Delete from `s.heartbeats`?

benbjohnson replied:
Good catch, fixed in  3384896.

titanous commented:
Looks like the commit needs a DCO sign-off before CI will run.

titanous commented:
This looks great!

josephglanville commented:
When trying to bootstrap with these changes I am seeing a ton of instance expiries:
https://gist.github.com/c1044cfe2cece0b87b38
https://gist.github.com/841a00a7072cec6fb47e

benbjohnson commented:
@titanous I split the commit into bolt upgrade (a3d873a) and heartbeating (3384896). I must have killed the DCO line when I rebased last time. I added the DCO to both commits.

benbjohnson commented:
@josephglanville `AddInstance()` wasn't returning an `ErrNotLeader` error for in-memory checks which was causing it to not redirect in the handler. Fixed here:

https://github.com/benbjohnson/flynn/commit/0c6f14ad9aacd28a67c8d459e551193349885d37#diff-e92f931b3419848b18a7209332fe268dR427

titanous commented:
The bolt update commit message should have a `Godeps: ` subsystem prefix.

benbjohnson commented:
@titanous I fixed the commit line. I'm just trying to figure out why the builds are failing now. It's showing 503s on different tests. Could it be caused by the longer leader election times from the 2x TTL?

titanous commented:
@benbjohnson I think that was a fluke or a timeout that is too short, we can fix after merging. I've got another run going.

titanous commented:
LGTM

thomasfedb commented:
When can I expect to see this change in a stable release?

titanous commented:
This weekend.

thomasfedb commented:
Excellent! :sparkles:
